pulser/sequence.py

```diff
@@ -282,27 +282,50 @@ def is_parametrized(self) -> bool:
         return not self._building
 
     @_screen
-    def get_duration(self, channel: Optional[str] = None) -> int:
+    def get_duration(
+        self, channel: Optional[str] = None, include_fall_time: bool = False
+    ) -> int:
         """Returns the current duration of a channel or the whole sequence.
 
         Keyword Args:
             channel (Optional[str]): A specific channel to return the duration
                 of. If left as None, it will return the duration of the whole
                 sequence.
+            include_fall_time (bool): Whether to include in the duration the
+                extra time needed by the last pulse to finish, if there is
+                modulation.
 
         Returns:
             int: The duration of the channel or sequence, in ns.
         """
         if channel is None:
-            durations = [
-                self._last(ch).tf
-                for ch in self._schedule
-                if self._schedule[ch]
-            ]
-            return 0 if not durations else max(durations)
+            channels = tuple(self._channels.keys())
+            if not channels:
+                return 0
+        else:
+            self._validate_channel(channel)
+            channels = (channel,)
+        last_ts = {}
+        for id in channels:
+            this_chobj = self._channels[id]
+            temp_tf = 0
+            for i, op in enumerate(self._schedule[id][::-1]):
+                if i == 0:
+                    # Start with the last slot found
+                    temp_tf = op.tf
+                    if not include_fall_time:
+                        break
+                if isinstance(op.type, Pulse):
+                    temp_tf = max(
+                        temp_tf, op.tf + op.type.fall_time(this_chobj)
+                    )
+                    break
+                elif temp_tf - op.tf >= 2 * this_chobj.rise_time:
+                    # No pulse behind 'op' with a long enough fall time
+                    break
```

Comment:
I'm not sure about that. The part you removed is there just to avoid the loop going over the entire list of instructions in the schedule, removing it just makes it go all the way to the end but it doesn't affect the result.